### PR TITLE
Adding unit tests for functions

### PR DESCRIPTION
**Pull request by thomik-corp**
- To test the functions:
  - Start the emulator from the parent folder via `npm run dev` 
  OR 
  - Run `npm run test:functions` (this automatically starts the emulator).
- These tests now run from CI/CD pipelines.
- Code coverage is published as artifacts in circleci.

### CODE REVIEW

File: functions/package.json [hunk 1]
```diff
@@ -7,7 +7,10 @@
     "shell": "npm run build && firebase functions:shell",
     "start": "npm run shell",
     "deploy": "firebase deploy --only functions",
-    "logs": "firebase functions:log"
+    "logs": "firebase functions:log",
+    "test": "set -a && . ./.testenv && jest --coverage  --detectOpenHandles && set +a",
```
Comment (rhelmer) [inline functions/package.json:11]:
```suggestion
    "test": "set -a && . ./.testenv && jest --coverage --detectOpenHandles && set +a",
```

Just an extra space

File: functions/src/__tests__/index.test.ts [hunk 2]
```diff
@@ -0,0 +1,248 @@
...
+    (response.status as jest.Mock).mockClear();
+    (fakeAuth.verifyIdToken as jest.Mock).mockClear();
+    (fakeAuth.createCustomToken as jest.Mock).mockClear();
+    send.mockClear();
+
+    await rallytoken(
+      {
+        method: "POST",
+        body: { idToken, studyId },
+      } as functions.Request<any>, // eslint-disable-line @typescript-eslint/no-explicit-any
```
Comment (rhelmer) [inline functions/src/__tests__/index.test.ts:242]: It'd be nice to declare our types, I'll file an issue, no need to worry about that in this PR.

**Review by rhelmer — CHANGES_REQUESTED** [review]
This looks great! The only other thing I'd add is that I noticed in the code coverage output it's not exercising deleting the "user studies" `user/{uid}/studies` subcollection:

```
------------|---------|----------|---------|---------|-------------------
File        | % Stmts | % Branch | % Funcs | % Lines | Uncovered Line #s 
------------|---------|----------|---------|---------|-------------------
All files   |   93.84 |     87.5 |     100 |   93.84 |                   
 index.ts   |   93.75 |     87.5 |     100 |   93.75 | 129-134           
 studies.ts |     100 |      100 |     100 |     100 |                   
------------|---------|----------|---------|---------|-------------------
```

Which is this bit:

https://github.com/mozilla-rally/rally-web-platform/blob/35279a3e9de674f6f5c2f50ecacba743c69fc568/functions/src/index.ts#L124-L142

This is a subcollection of `/users/{uid}` and is used to store the list of studies that the user has consented to, for instance if I join the `facebookPixelHunt` study then I would have a document in `/users/{uid}/studies/facebookPixelHunt`. Both the website and the extension watch this document to see if consent has been recorded for this study.

You can see where the website modifies this: https://github.com/mozilla-rally/rally-web-platform/blob/35279a3e9de674f6f5c2f50ecacba743c69fc568/src/lib/stores/api.js#L70-L73

Unfortunately, Firebase cannot recursively delete subcollections, and it must be done in batches in case the number of documents gets large in the future, so having test coverage here would be especially helpful :)

Would you mind adding a test for this? Ideal would be to add a number of documents over the batch limit (it works in batches of 4) to ensure that there isn't a bug introduced here.

The consequence of getting this wrong isn't too dire, but it does mean useless garbage builds up in Firestore which we'd like to avoid, and we also want to honor the user's request to delete all of their data.

File: functions/src/__tests__/index.test.ts (hunk 2 above)
Comment (thomik-corp) [reply functions/src/__tests__/index.test.ts:242]: Agreed !

**Review by rhelmer — APPROVED** [review]

**Pull request merged**